cmake/OpenCogFunctions.cmake

```diff
@@ -0,0 +1,117 @@
+# Copyright (C) 2016 OpenCog Foundation
+
+# References:
+# https://www.gnu.org/software/guile/manual/guile.html#Modules-and-the-File-System
+# https://www.gnu.org/software/guile/manual/guile.html#Creating-Guile-Modules
+# https://www.gnu.org/software/guile/manual/guile.html#Installing-Site-Packages
+
+# Definitions:
+#
+# * MODULE_FILE: The name of the file that defines the module. It has
+#   the same name as the directory it is in, or the name of the parent
+#   directory of current directory if it is in a folder named 'scm'.
+#   In addition this file shoule have a define-module expression
+#   for it be importable, as per guile's specification. See reference
+#   links above.
+
+# ----------------------------------------------------------------------------
+# This configures the install and symlink paths for each file, passed to it,
+# based on the value of the variables MODULE_NAME, MODULE_FILE_DIR_PATH and
+# MODULE_DIR_PATH in the PARENT_SCOPE.
+FUNCTION(PROCESS_MODULE_STRUCTURE FILE_NAME)
+    SET(GUILE_SYMLINK_DIR "${CMAKE_BINARY_DIR}/opencog/scm")
+    SET(GUILE_INSTALL_DIR "${DATADIR}/scm")
+
+    # Create symlinks in build directory mirroring the install path structure.
+    # Also configure for install.
+    IF ("${MODULE_NAME}.scm" STREQUAL "${FILE_NAME}")
+        EXECUTE_PROCESS(
+            COMMAND ${CMAKE_COMMAND} -E make_directory ${GUILE_SYMLINK_DIR}/${MODULE_FILE_DIR_PATH}
+            COMMAND ${CMAKE_COMMAND} -E create_symlink "${CMAKE_CURRENT_SOURCE_DIR}/${FILE_NAME}" "${GUILE_SYMLINK_DIR}/${MODULE_FILE_DIR_PATH}/${FILE_NAME}"
```

Comment:
It is not this PR issue but rather an issue that exists in opencog/atomspace master branch.

EXECUTE_PROCESS chaining make_directory and create_symlink actions fails sporadically because directory is not created at the moment when symlinks is being created:
```
failed to create symbolic link '/tmp/test/atomspace-build/opencog/scm//opencog/matrix.scm': No such file or directory
```
Can be fixed by replacing one EXECUTE_PROCESS with two commands by two EXECUTE_PROCESS lines with one command in each:

    IF ("${MODULE_NAME}.scm" STREQUAL "${FILE_NAME}")
        EXECUTE_PROCESS(
            COMMAND ${CMAKE_COMMAND} -E make_directory ${GUILE_SYMLINK_DIR}/${MODULE_FILE_DIR_PATH}
        )
        EXECUTE_PROCESS(
            COMMAND ${CMAKE_COMMAND} -E create_symlink "${CMAKE_CURRENT_SOURCE_DIR}/${FILE_NAME}" "${GUILE_SYMLINK_DIR}/${MODULE_FILE_DIR_PATH}/${FILE_NAME}"
        )
        SET(FILE_INSTALL_PATH "${GUILE_INSTALL_DIR}/${MODULE_FILE_DIR_PATH}"
            PARENT_SCOPE
        )
    ELSE()
        EXECUTE_PROCESS(
            COMMAND ${CMAKE_COMMAND} -E make_directory ${GUILE_SYMLINK_DIR}/${MODULE_DIR_PATH}
        )
        EXECUTE_PROCESS(
            COMMAND ${CMAKE_COMMAND} -E create_symlink "${CMAKE_CURRENT_SOURCE_DIR}/${FILE_NAME}" "${GUILE_SYMLINK_DIR}/${MODULE_DIR_PATH}/${FILE_NAME}"
        )
        SET(FILE_INSTALL_PATH "${GUILE_INSTALL_DIR}/${MODULE_DIR_PATH}"
            PARENT_SCOPE
        )
    ENDIF()

Raising it here because it should be fixed either in this PR or after this PR is merged.